cloud/scope/powervs_cluster.go

```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+		return nil, fmt.Errorf("failed to create vpc, error getting id for resource group %s, %w", *s.ResourceGroup(), err)
+	}
+	addressPrefixManagement := "auto"
+	vpcOption := &vpcv1.CreateVPCOptions{
+		ResourceGroup:           &vpcv1.ResourceGroupIdentity{ID: &resourceGroupID},
+		Name:                    s.GetServiceName(VPC),
+		AddressPrefixManagement: &addressPrefixManagement,
+	}
+	vpcDetails, _, err := s.IBMVPCClient.CreateVPC(vpcOption)
+	if err != nil {
+		return nil, err
+	}
+
+	// set security group for vpc
+	options := &vpcv1.CreateSecurityGroupRuleOptions{}
+	options.SetSecurityGroupID(*vpcDetails.DefaultSecurityGroup.ID)
+	options.SetSecurityGroupRulePrototype(&vpcv1.SecurityGroupRulePrototype{
+		Direction: core.StringPtr("inbound"),
+		Protocol:  core.StringPtr("tcp"),
+		IPVersion: core.StringPtr("ipv4"),
+		PortMin:   core.Int64Ptr(int64(s.APIServerPort())),
+		PortMax:   core.Int64Ptr(int64(s.APIServerPort())),
+	})
+	_, _, err = s.IBMVPCClient.CreateSecurityGroupRule(options)
+	if err != nil {
+		return nil, err
+	}
+	return vpcDetails.ID, nil
+}
+
+// ReconcileVPCSubnet reconciles VPC subnet.
+func (s *PowerVSClusterScope) ReconcileVPCSubnet() error {
+	//TODO: Should we handle a case where there are no subnet specified
+	for _, subnet := range s.IBMPowerVSCluster.Spec.VPCSubnets {
+		if subnet.Name == nil {
+			subnet.Name = pointer.String(fmt.Sprintf("%s-vpcsubnet", s.InfraCluster().GetName()))
+		}
+		s.Info("Reconciling vpc subnet", "subnet", subnet.Name)
+		subnetID := s.GetVPCSubnetID(subnet)
+		if subnetID != nil {
+			subnetDetails, _, err := s.IBMVPCClient.GetSubnet(&vpcv1.GetSubnetOptions{
+				ID: subnetID,
+			})
+			if err != nil {
+				return err
+			}
+			if subnetDetails == nil {
+				return fmt.Errorf("error failed to get vpc subnet with id %s", *subnetID)
+			}
+			// check for next subnet
+			continue
+		}
+
+		// check VPC subnet exist in cloud
+		vpcSubnetID, err := s.checkVPCSubnet(subnet)
+		if err != nil {
+			s.Error(err, "error checking vpc subnet")
+			return err
+		}
+		if vpcSubnetID != "" {
+			s.Info("found vpc subnet", "id", vpcSubnetID)
+			s.SetVPCSubnetID(*subnet.Name, infrav1beta2.ResourceReference{ID: &vpcSubnetID, ControllerCreated: pointer.Bool(false)})
+			// check for next subnet
+			continue
+		}
+		subnetID, err = s.createVPCSubnet(subnet)
+		if err != nil {
+			s.Error(err, "error creating vpc subnet")
+			return err
+		}
+		if subnetID != nil {
+			s.Info("created vpc subnet", "id", subnetID)
+			s.SetVPCSubnetID(*subnet.Name, infrav1beta2.ResourceReference{ID: subnetID, ControllerCreated: pointer.Bool(true)})
+		}
+		// TODO(karthik-k-n)(Doubt): Do we need to create public gateway?
+	}
+	return nil
+}
+
+// checkVPCSubnet checks VPC subnet exist in cloud.
+func (s *PowerVSClusterScope) checkVPCSubnet(subnet infrav1beta2.Subnet) (string, error) {
+	vpcSubnet, err := s.IBMVPCClient.GetVPCSubnetByName(*subnet.Name)
+	if err != nil {
+		return "", err
+	}
+	if vpcSubnet == nil {
```

Comment:
Not required like dhcp get

Reply:
I think its required as vpcsubnet may not be found in cloud